### PR TITLE
Ace3 / String Exporter

### PR DESCRIPTION
**Pull request by zdeyn**
I've put together a demo of the Ace3 String Exporter:

Using this PR, type `/hc ace` in-game to see what I've been talking about.
You'll note the string listed is really.. really.. really long - copy-paste it into notepad to see just how big.

We can move a LOT of data.

Once I'm sure people are happy with this, I'll switch over the button on the standard Hardcore window to use this version, and push on with other awesome

### CODE REVIEW

**Comment by RobertB4** [comment]
Can you make it work within the current addon UI? Meaning that when you click the "Get verified" button, it shows inside there. That's the part I was having trouble with.

**Comment by zdeyn** [comment]
> Can you make it work within the current addon UI? Meaning that when you click the "Get verified" button, it shows inside there. That's the part I was having trouble with.

I can make it show instead of 'there' e.g. when clicking the 'get verified' button, it doesn't go anywhere anymore, just pops up the new box.

My plan is to eventually replace all of the interface with Ace3-foundation Frames, which we can then restyle and present as we wish, within limits - this won't take as long as one might think.

Do I have permission to replace the 'get verified' button with this dialog box?

**Comment by RobertB4** [comment]
> I can make it show instead of 'there' e.g. when clicking the 'get verified' button, it doesn't go anywhere anymore, just pops up the new box.

Oh, I must have misunderstood you then, I thought you planned on implementing it inside our current UI.

I think there is value in making a new Frame with Ace3 and ditching the current UI, but probably not before release.
If you are up for the task, implement the string exporter with Ace3 and take a screenshot of what it looks like, so we can run it past Brandung and Kargoz to get input from them.

**Comment by zdeyn** [comment]
image

Here's the current version - it's shown by:
- a) using the command `/hardcore verify`, or
- b) clicking the "Get verified" button the main window, as shown in the screenshot above

There's still a few minor things I'm sorting out (such as preventing a 2nd window from opening when one is already open) but that's only a few minutes away from done

Meanwhile, here's the requested screenshot - I look forward to the feedback

**Comment by zdeyn** [comment]
By the way, this version uses an export string which is double the length of the default created - and that string is doubled again another 10 times... so, 1024 times larger than the previous export string being displayed.

We've got a lot more room to send data via copy-paste than you suspect

**Comment by zdeyn** [comment]
single instance window fixed

**Review by blenderer — COMMENTED** [review]
Are all of the Ace3 imports necessary for this 1 window to work? Not a huge deal just wondering.

Please rebase on main as soon as possible. LMK if you need help with that

**Comment by zdeyn** [comment]
Its a full installation which event based - if you don't write to call it, it'd not even load (LibSt7b)

**Comment by zdeyn** [comment]
Sorry. Fat fingrred it

**Review by blenderer — COMMENTED** [review]
Demo is cool, I was already sold. What's the timeline for an implementation when we hit the Verify button?

**Comment by zdeyn** [comment]
It's already in the PR now

File: Hardcore.lua
Comment (aaronma37) [inline Hardcore.lua:178]: This is just for demo sake right?

Comment (zdeyn) [reply Hardcore.lua:178]: You can merge to main or run my branch to test, but the feature is complete and done as a demo, and only fix for going live is removing the loop which doubles the export string (for the demo)

Comment (aaronma37) [reply Hardcore.lua:178]: Yeah thats what I meant - the for loop + fake string stuff. Sweet

**Comment by zdeyn** [comment]
image

The current look-and-feel. The sample-code generator has been removed, and the box now presents an authentic code to the player.

The box can be moved and resized.

File: Hardcore.lua
Comment (RobertB4) [inline Hardcore.lua:155]: this variable is unused and can be removed safely.

File: Hardcore.lua
Comment (RobertB4) [inline Hardcore.lua:147]: This state feels weird. Is there a specific reason you did not use the Show and Hide functions of the frame itself?

```
widget:Show()
widget:Hide()
```

**Review by RobertB4 — COMMENTED** [review]
All the code related to the Get Verified tab can be removed, there is a bunch we no longer need.

**Comment by RobertB4** [comment]
I think there is value in having this in the initial release, let's make sure we get it ready asap!

File: Hardcore.lua
Comment (zdeyn) [reply Hardcore.lua:147]: > This state feels weird. Is there a specific reason you did not use the Show and Hide functions of the frame itself?
> 
> ```
> widget:Show()
> widget:Hide()
> ```

Other than being being a noob? None that come to mind - If you have a quick-fix then be my guest!

My intent was to ensure that there would only ever be one Ace3 Window Dialog available (for this feature) at once. I missed the mark. I suspect there's a problem with multiple windows if the slash command is used while a window is already showing - I haven't tested it.

"Get it perfect now", "Get it out the door now, fix it later" ?

**Comment by zdeyn** [comment]
> I think there is value in having this in the initial release, let's make sure we get it ready asap!

It could ship as-is, but it's a little unstable - I'll do what I can, and we can pull the rip-cord whenever

File: Hardcore.lua
Comment (zdeyn) [reply Hardcore.lua:147]: > This state feels weird. Is there a specific reason you did not use the Show and Hide functions of the frame itself?

Turns out, there is:
```lua
function Hardcore:CloseVerifiedDialog(widget)
	Hardcore.verify_shown = false
	AceGUI:Release(widget)
end```

Ace3 is recycling the dialog's resources for use elsewhere when it's not shown.. it's not 'hiding it', it's destroying it.

**Comment by zdeyn** [comment]
> Please rebase on main as soon as possible. LMK if you need help with that

Yep, I need to be shown how to handle a merge like this properly - please let me know when you're free. Worst-case, just perform the rituals and say the words on your own, and I'll learn from the next one!

Appreciate the assist

- Z.

File: Hardcore.lua
Comment (RobertB4) [reply Hardcore.lua:147]: You are creating a new widget every time call `Hardcore:GetVerifiedDialog` which I assume is the problem.

You could make the dialog a singleton and use :Hide() and :Show() to toggle its visibility

I haven't tested if the code works, but something like:
```
GetVerifiedDialog = nil -- this should be in the global scope

function Hardcore:CreateGetVerifiedDialog() -- call this function once when the addon is initialized so the dialog is ready when needed
 -- create widget with ace 3 in here
end
```

And then toggle the dialog whenever needed, e.g. when clicking the Get Verified button defined in the xml file:

```
<OnClick>
    GetVerifiedDialog:Show()
</OnClick>
```

**Comment by RobertB4** [comment]
> It could ship as-is, but it's a little unstable - I'll do what I can, and we can pull the rip-cord whenever

Can you be a bit more specific what you mean by unstable? If it is working as intended, I think it is ok to merge it and clean up the code later, but if there are bugs, we should fix them before merging.

File: Hardcore.lua
Comment (zdeyn) [reply Hardcore.lua:147]: > You are creating a new widget every time call `Hardcore:GetVerifiedDialog` which I assume is the problem.
> 
> You could make the dialog a singleton and use :Hide() and :Show() to toggle its visibility

That would be the pattern in GUIs other than Ace3, however in WoW you can't ever destroy something - once created, it's always there until the GC is forced to handle it.

Ace3 overcomes this by using a fixed set of resources, and restyling/repositioning them for each different dialog you view: the heading frame and close box on the String Exporter is the same as used in another Ace3 addon's UI - there's 'enough' if you want to run multiple Ace3 apps and have them displayed at the same time, but there's still only a finite number of them.

The pattern used above allows Ace3, once the window is 'hidden' to reallocate the resources for use on other dialogs. Our dialog ceases to exist - it's not actually 'hidden', but destroyed.

Thus, when trying to show multiple instances of this window, we need to avoid RE-creating it, if it's already there and showing - that would work, but it would occupy more 'can never be reclaimed' slots (one per widget component)

So, yes - the pattern is strange, but I understand it - the code as committed is correct.

Comment (RobertB4) [reply Hardcore.lua:147]: > as it's not 'Show/Hide' but 'Build/Destroy'

Isn't that the same thing?
Build the widget when needed (e.g. when clicking the Get Verified button)
Destroy the widget when no longer needed (when closing the window)

Either way there shouldn't be the need to have external state to display/hide (create/destroy) a widget.

**Comment by zdeyn** [comment]
> Can you be a bit more specific what you mean by unstable? If it is working as intended, I think it is ok to merge it and clean up the code later, but if there are bugs, we should fix them before merging.

I've had a poke and cleaned up a few things, I think it's good-to-go, but could use some other eyes and players on it just to make sure - there's a lot of files being pushed here.

**Comment by RobertB4** [comment]
I will test this later today. If it is working, I think we should merge it.

**Review by RobertB4 — APPROVED** [review]
Tested on SoM beta, it's working as intended.

**Comment by zdeyn** [comment]
Ok, so how do I/we merge this in properly? Can someone either tell me, show me, or just do it and I'll learn it later?

File: Hardcore.lua
Comment (blenderer) [inline Hardcore.lua:153]: Can we move this to top of file?

**Review by blenderer — COMMENTED** [review]
Had 1 comment about ace intialization line, then good to merge after

File: Hardcore.lua
Comment (zdeyn) [reply Hardcore.lua:153]: It can only be executed after the `Hardcore` Frame is created, on line 145:

If we wanted to move everything to the start of the file, we can do that, so long as we have everything we need already init'd before we create `Hardcore`

**Comment by zdeyn** [comment]
Posted a re-arranged version - does this fit your needs? `Hardcore` Frame object and Ace3 bits'n'pieces are both loaded closer to the top of the file - just need to make sure all dependency objects are created first, in the future

**Pull request closed without merging**
